# 4.9 TASK Part 1 

#### Importing Libraries and Data

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
path = r'C:\Users\ophab\Desktop\JESSICA\Careerfoundry JEJE\Data Immersion\09-2025 Instacart Basket Analysis'

In [4]:
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [5]:
customers.shape

(206209, 10)

In [7]:
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [10]:
customers.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [11]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


## Data Wrangling & Consistency Checks

### Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

### Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

### Dropping columns

In [12]:
customers.drop(columns=['First Name', 'Surnam'], inplace=True)

In [13]:
customers.head()

user_id  Gender       STATE  Age date_joined  n_dependants fam_status  \
0    26711  Female    Missouri   48    1/1/2017             3    married   
1    33890  Female  New Mexico   36    1/1/2017             0     single   
2    65803    Male       Idaho   35    1/1/2017             2    married   
3   125935  Female        Iowa   40    1/1/2017             0     single   
4   130797  Female    Maryland   26    1/1/2017             1    married   

   income  
0  165665  
1   59285  
2   99568  
3   42049  
4   40374

In [14]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   Gender        206209 non-null  object
 2   STATE         206209 non-null  object
 3   Age           206209 non-null  int64 
 4   date_joined   206209 non-null  object
 5   n_dependants  206209 non-null  int64 
 6   fam_status    206209 non-null  object
 7   income        206209 non-null  int64 
dtypes: int64(4), object(4)
memory usage: 12.6+ MB


##### I decided to drop the First Name and Surname columns from the customers dataframe because these columns contain missing values and are not relevant for the analysis. The focus of this project is on customer demographics, behavior, and spending patterns, so personally identifying information like names is not needed. Dropping these columns simplifies the dataframe and ensures we only keep the data that contributes to the analytical objectives.

### Missing Values

In [20]:
customers.isnull().sum()

user_id         0
gender          0
state           0
age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

##### No missing values found.

### Renaming Columns

In [16]:
customers.rename(columns={'Gender': 'gender','STATE': 'state','Age': 'age',}, inplace=True)

In [17]:
customers.head()

user_id  gender       state  age date_joined  n_dependants fam_status  \
0    26711  Female    Missouri   48    1/1/2017             3    married   
1    33890  Female  New Mexico   36    1/1/2017             0     single   
2    65803    Male       Idaho   35    1/1/2017             2    married   
3   125935  Female        Iowa   40    1/1/2017             0     single   
4   130797  Female    Maryland   26    1/1/2017             1    married   

   income  
0  165665  
1   59285  
2   99568  
3   42049  
4   40374

##### I renamed the columns Gender, STATE, and Age to gender, state, and age respectively to standardize the column names. Using lowercase and consistent naming conventions improves readability, makes coding easier, and ensures consistency throughout the analysis. This step does not change the data itself, only the labels for clarity and better workflow management.

### Changing Data Type

In [18]:
customers['user_id'] = customers['user_id'].astype(str)

In [19]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  object
 1   gender        206209 non-null  object
 2   state         206209 non-null  object
 3   age           206209 non-null  int64 
 4   date_joined   206209 non-null  object
 5   n_dependants  206209 non-null  int64 
 6   fam_status    206209 non-null  object
 7   income        206209 non-null  int64 
dtypes: int64(3), object(5)
memory usage: 12.6+ MB


##### I converted the user_id column from integer to string to ensure consistency across all dataframes. This simplifies merging operations and prevents potential type mismatch issues when combining customer data with other datasets. The conversion does not alter the actual IDs, only their data type for easier handling.

### Investigating Mixed-Type Columns

In [22]:
for col in customers.columns.tolist():
  weird = (customers[[col]].map(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

##### None found

### Duplicates

In [23]:
customers[customers.duplicated(subset=['user_id'], keep=False)]

Empty DataFrame
Columns: [user_id, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

##### I checked for duplicates in the user_id column because it serves as the unique identifier for each customer, while other columns like gender, state, or age can naturally have repeated values. No duplicates were found, which confirms that each customer has a unique ID. This ensures accurate merging with other datasets and reliable calculation of customer-level metrics, avoiding potential errors caused by duplicate IDs.

## End of Data Wrangling & COnsistency Checks

## Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [24]:
# Import dataframe for merging
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_grouping_REVISED.pkl'))

In [26]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   Unnamed: 0_x            int64   
 1   order_id                int64   
 2   user_id                 int64   
 3   eval_set                object  
 4   order_number            int64   
 5   orders_day_of_week      int64   
 6   order_hour_of_day       int64   
 7   days_since_prior_order  float64 
 8   product_id              int64   
 9   add_to_cart_order       int64   
 10  reordered               int64   
 11  _merge                  category
 12  Unnamed: 0_y            int64   
 13  product_name            object  
 14  aisle_id                int64   
 15  department_id           int64   
 16  prices                  float64 
 17  merge_flag              category
 18  price_range             object  
 19  busiest_day             object  
 20  day_category            object  
 21  busies

### Wrangling and Consistency on New Imported Dataframe 

In [28]:
# Dropping unwanted columns
cols_to_drop = ['Unnamed: 0_x', 'eval_set', '_merge', 'Unnamed: 0_y', 'merge_flag']
ords_prods_merge.drop(columns=cols_to_drop, inplace=True)

In [29]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 20 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
 6   product_id              int64  
 7   add_to_cart_order       int64  
 8   reordered               int64  
 9   product_name            object 
 10  aisle_id                int64  
 11  department_id           int64  
 12  prices                  float64
 13  price_range             object 
 14  busiest_day             object 
 15  day_category            object 
 16  busiest_period_of_day   object 
 17  loyalty_flag            object 
 18  max_order               int64  
 19  order_frequency_flag    object 
dtypes: float64(2), int64(11), object(7)
memory usage: 4.8+ GB


In [30]:
# Convert Datatypes
ords_prods_merge['order_id'] = ords_prods_merge['order_id'].astype(str)
ords_prods_merge['user_id'] = ords_prods_merge['user_id'].astype(str)
ords_prods_merge['product_id'] = ords_prods_merge['product_id'].astype(str)

In [31]:
# Check numeric columns for consistency
ords_prods_merge[['order_number','orders_day_of_week','order_hour_of_day','days_since_prior_order',
                  'add_to_cart_order','reordered','aisle_id','department_id','prices','max_order']].describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.240486e+07        3.240486e+07       3.240486e+07   
mean   1.714230e+01        2.738867e+00       1.342515e+01   
std    1.753532e+01        2.090077e+00       4.246380e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.400000e+01        5.000000e+00       1.600000e+01   
max    9.900000e+01        6.000000e+00       2.300000e+01   

       days_since_prior_order  add_to_cart_order     reordered      aisle_id  \
count            3.032876e+07       3.240486e+07  3.240486e+07  3.240486e+07   
mean             1.110408e+01       8.352547e+00  5.895873e-01  7.119612e+01   
std              8.779064e+00       7.127071e+00  4.919087e-01  3.821139e+01   
min              0.000000e+00       1.000000e+00  0.000000e+00  1.000000e+00   
25%              5.000000e+00       3.000000e+00  0.000000e+00  3.100000e+01   
50%              8.000000e+00       6.000000e+00  1.000000e+00  8.300000e+01   
75%              1.500000e+01       1.100000e+01  1.000000e+00  1.070000e+02   
max              3.000000e+01       1.450000e+02  1.000000e+00  1.340000e+02   

       department_id        prices     max_order  
count   3.240486e+07  3.239973e+07  3.240486e+07  
mean    9.919792e+00  7.790994e+00  3.305217e+01  
std     6.281485e+00  4.241809e+00  2.515525e+01  
min     1.000000e+00  1.000000e+00  1.000000e+00  
25%     4.000000e+00  4.200000e+00  1.300000e+01  
50%     9.000000e+00  7.400000e+00  2.600000e+01  
75%     1.600000e+01  1.130000e+01  4.700000e+01  
max     2.100000e+01  2.500000e+01  9.900000e+01

In [33]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 20 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
 6   product_id              object 
 7   add_to_cart_order       int64  
 8   reordered               int64  
 9   product_name            object 
 10  aisle_id                int64  
 11  department_id           int64  
 12  prices                  float64
 13  price_range             object 
 14  busiest_day             object 
 15  day_category            object 
 16  busiest_period_of_day   object 
 17  loyalty_flag            object 
 18  max_order               int64  
 19  order_frequency_flag    object 
dtypes: float64(2), int64(8), object(10)
memory usage: 4.8+ GB


In [34]:
# Convert numeric IDs to category 
ords_prods_merge['aisle_id'] = ords_prods_merge['aisle_id'].astype('category')
ords_prods_merge['department_id'] = ords_prods_merge['department_id'].astype('category')

In [35]:
# Convert day/hour to category for grouping
ords_prods_merge['orders_day_of_week'] = ords_prods_merge['orders_day_of_week'].astype('category')
ords_prods_merge['order_hour_of_day'] = ords_prods_merge['order_hour_of_day'].astype('category')

In [36]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 20 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                object  
 1   user_id                 object  
 2   order_number            int64   
 3   orders_day_of_week      category
 4   order_hour_of_day       category
 5   days_since_prior_order  float64 
 6   product_id              object  
 7   add_to_cart_order       int64   
 8   reordered               int64   
 9   product_name            object  
 10  aisle_id                category
 11  department_id           category
 12  prices                  float64 
 13  price_range             object  
 14  busiest_day             object  
 15  day_category            object  
 16  busiest_period_of_day   object  
 17  loyalty_flag            object  
 18  max_order               int64   
 19  order_frequency_flag    object  
dtypes: category(4), float64(2), int64(4), object

In [37]:
ords_prods_merge.isnull().sum()

order_id                        0
user_id                         0
order_number                    0
orders_day_of_week              0
order_hour_of_day               0
days_since_prior_order    2076096
product_id                      0
add_to_cart_order               0
reordered                       0
product_name                    0
aisle_id                        0
department_id                   0
prices                       5127
price_range                     0
busiest_day                     0
day_category                    0
busiest_period_of_day           0
loyalty_flag                    0
max_order                       0
order_frequency_flag            5
dtype: int64

In [38]:
ords_prods_merge['days_since_prior_order'].isnull().sum()

np.int64(2076096)

##### The column days_since_prior_order contains a large number of missing values, which correspond to first orders for each customer. Since there is no previous order to calculate the difference from, these NaN values are meaningful and indicate the start of a customer’s ordering history. Following the tutor’s suggestion, the NaN values were left as-is rather than filled with 0 or another value. This approach preserves the original information, allows accurate calculation of metrics like average days between orders, and avoids introducing artificial bias into the dataset.

In [39]:
# Fill missing prices with the median
median_price = ords_prods_merge['prices'].median()
ords_prods_merge['prices'].fillna(median_price, inplace=True)

C:\Users\ophab\AppData\Local\Temp\ipykernel_3764\3932637507.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  ords_prods_merge['prices'].fillna(median_price, inplace=True)


In [40]:
ords_prods_merge['prices'].isnull().sum()

np.int64(0)

##### The prices column had a few missing values (5,127). These were filled with the median price of the dataset. Using the median preserves the overall distribution and avoids distortion from extreme price values, while preventing loss of rows for analysis.

In [41]:
ords_prods_merge['order_frequency_flag'].isnull().sum()

np.int64(5)

##### The order_frequency_flag column contains a very small number of missing values (5). These were left as NaN to accurately reflect that the customer’s ordering frequency could not be determined. Leaving them ensures that no artificial information is introduced and that analyses correctly account for unknown cases.

## Merging Dataframes

In [42]:
customers['user_id'].nunique()

206209

In [43]:
ords_prods_merge['user_id'].nunique()


206209

In [45]:
customers['user_id'].dtype

dtype('O')

In [46]:
ords_prods_merge['user_id'].dtype

dtype('O')

In [47]:
# Merge orders-products dataset with customers dataset
ords_prods_cust = ords_prods_merge.merge(customers, on='user_id', how='left')

In [48]:
# Quick checks              

In [49]:
ords_prods_cust.shape 

(32404859, 27)

In [50]:
ords_prods_cust[['user_id','order_id']].nunique()

user_id      206209
order_id    3214685
dtype: int64

In [51]:
ords_prods_cust.isnull().sum() 

order_id                        0
user_id                         0
order_number                    0
orders_day_of_week              0
order_hour_of_day               0
days_since_prior_order    2076096
product_id                      0
add_to_cart_order               0
reordered                       0
product_name                    0
aisle_id                        0
department_id                   0
prices                          0
price_range                     0
busiest_day                     0
day_category                    0
busiest_period_of_day           0
loyalty_flag                    0
max_order                       0
order_frequency_flag            5
gender                          0
state                           0
age                             0
date_joined                     0
n_dependants                    0
fam_status                      0
income                          0
dtype: int64

##### The merged dataset contains 206,209 unique customers and 3,214,685 orders, confirming that all users from the customers dataset are represented and the merge was successful.

In [54]:
ords_prods_cust.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_cust_merged.pkl'))